### DISCUSSION TITLE
Parse json containing an object with unknown keys as a sorted vector of tables

### DISCUSSION THREAD

**Discussion by trws** [dvidelabs/flatcc#291] [Q&A]
I've been very impressed with flatcc so far, but seem to have run into a challenge I can't seem to work my way out of.  I have an application that makes heavy use of json, where we would really like to switch to a more structured schema-oriented format incrementally, and have the option to go binary later.  It's mainly C with some C++, so flatcc looks like a fantastic fit.  The trick is, we have a _lot_ of existing code and files with pre-generated json we need to keep supporting.  I've been doing some tests with one of the more horrendous examples, a 400+MB json document, a very small chunk of which looks like this:

```json
{
    "R_lite": [
      {
        "rank": "0",
        "children": {
          "core": "0-23"
        }
      },
      {
        "rank": "1-60",
        "children": {
          "core": "0-15"
        }
      }
    ]
}
```

The "children" key of each entry under "R_lite" is a generic map of string to string where the key is a resource type and the value is an IDset (range-encoded set of ints separated by commas).   There are a few things like this in our json documents, and I can't seem to work out how to make a schema that can parse these without hitting "expected array" or similar.  IIRC they could be handled as "unknown fields" but that wouldn't get me the data I need.  Here's what I've been trying:

```
table pair
{
key:string (required, key);
val:string;
}
table Rlite
{
rank:  string;
children:  [pair] (sorted);
}
```

Hence the "expected array" error.  Is there a way to mark in the schema that a vector of tables should be parsed/generated as an object rather than a json array, or some other way to parse this json with jsoncc?

**Comment by mikkelfj** [comment]
Thanks for the positive feedback.
There isn't really a way to do what you want with the JSON syntax you present, except by using something like jquery to translate the data to a new syntax before piping the data into the FlatCC JSON parser, but that might also work rather well in praxis.

From memory without testing, you can create a new syntax using your suggested schema:

```
{
    "R_lite": [
      {
        "rank": "0",
        "children": [
          { "key": "core", "val": "0-23" }
        ]
      },
      {
        "rank": "1-60",
        "children": [
          { "key": "core", "val": "0-15" }
        ]
      }
    ]
}
```

You can add a "key" attribute to the keys field in the schema as you have already done, which would let flatbuffers in general allow for sorting and binary searching. FlatCC specifically also supports the sorted attribute (as you also use), which will cause it to generate a sort function that will traverse the entire buffer and sort all fields marked as sorted. You can also call the sort operation on specific fields individually without the sorted attribute. That has to do with allowing a table with a key attribute to be used both in sorted and unsorted tables. Note that FlatCC does not automatically sort during construction for technical and performance reasons, but does it inline on a newly created buffer if asked to do so. FlatCC also supports unsorted search through linear scan methods.

A more FlatBuffer native version, but less JSON native structure would separate key and value arrays. Here you would search the key, and use the resulting index to find the value. This approach is much more efficient since in does not need to construct an object for each item, but either way should work:

```

table Rlite
{
rank:  string;
childkeys: [string];
childvals: [string];
}
```
I have not shown the corresponding JSON, but I guess you have the idea by now, or please ask again.

Note that in this case there are neither sorted or key attributes, but you should still be able to sort. I guess you could add sorted attribtute to the childkeys field, but it has been a while since I worked with this.

Incidentally, your approach, or my suggested alternative, is what I recommend when you dealing arbitrary key value data. This is one reason FlatCC does not support FlexBuffers that are available in Googles C++ project. I feel that JSON works just as well as any other format when you have odd data, but then FlatCC has a comparatively faster JSON to FlatBuffers parser. That said, you might also want to look into FlexBuffers, but it probably won't help you with parsing JSON directly, and it will be slower.

Side note:
I have been working on and off on a separate JSON superset configuration language that can understand common JSON patterns, and potentially parse them into FlatBuffers, but it is sort of low priority slow work, and it would not be portable to FlatBuffers tooling, except as data in form of flatbuffers. A use case could be parsing Open Street Map data.

**Reply by trws** [reply]
For some of our uses I can probably change the format right off, at least where we control both ends of the channel.  FWIW I actually tried FlatBuffer with flexbuffer, which handles this json parsing case just fine. The downfall is that the json handling is _SLOW_ in flatbuffers/flatc. Slow enough that only nlohmann::json and protobuf of the c++ parsers were slower.

I suppose I was hoping there might be an attribute or something that would let us treat that sorted vector as an object for the purpose of json. It would be a bit like flexbuffers I suppose, but much more constrained, just saying "this vector is a set of pair tables pulled from an object with the key from whatever is marked (key) and other field as value specifically and only for json."  That way a document like this could be handled without a re-structuring in json but use the native vector layout in an actual flatbuffer.  Almost want to try implementing it and see what happens but I'm not 100% sure where to start with adding a custom attribute.

**Reply by mikkelfj** [reply]
The custom attributes part is not very friendly to backends as it is because attributes are presented as a parse tree, not a map. Most attributes are handled by detecting standard attributes early in the compiler and setting a flag. If this were to be a standard feature of the JSON, and if there are already similar flags and space in the flag space, that could be way to deal with it. More generally a PR that cleans up attribute access would be good. I cannot afford too much time on this, but I can off guidance.

More generally, the problem with parsing JSON is that there are so many different "reasonable" patterns, so choosing one to integrate for a specific use case is a bit concerning, especially when it is not compatible with the C++ JSON parser. However, if you say that it is working with flexbuffers that would be a starting point. I would probably have to discuss it with other stakeholders, and again, it would be a PR where I cannot invest too much time into it. That said, generic attribute sets in JSON could be useful.

Another example of  a useful JSON pattern is one where a single value is converted into an array if the underlying field type is an array. I already did a config language prototype language doing that as a flatcc backend, but I have aborted it in favour of a different more flexible language since it became too difficult to formalize.

Also be warned that it is not easy to work on the JSON parser code generator. I'm not entirely happy with how complex it is, due to many special cases, of which you are about to add more, but it is difficult to avoid if you don't want to compromise on output quality or by moving complexity elsewhere.

**Reply by trws** [reply]
> More generally, the problem with parsing JSON is that there are so many different "reasonable" patterns, so choosing one to integrate for a specific use case is a bit concerning, especially when it is not compatible with the C++ JSON parser. However, if you say that it is working with flexbuffers that would be a starting point. I would probably have to discuss it with other stakeholders, and again, it would be a PR where I cannot invest too much time into it. That said, generic attribute sets in JSON could be useful.

Yeah, I should probably elaborate on what I tried and what seems to work today.  Using google/flatbuffers c++ (looks like not all bindings support this yet), a schema like this:

```
table Rlite
{
rank:  string;
children: [ubyte] (flexbuffer);
}
```

will read the `children` element of the json I originally posted and make it available as a flexbuffers map.  The thing that's unfortunate about this is that in binary it could be any flexbuffer data, and it doesn't have a clear mapping into the flatcc API as far as I can tell.  My original thought was that it would be natural to make it so that an attribute, like jsonobject or something 🤷, could be used to map something like `children: [pair] (jsonobject)` to a vector of (in the simple case) pairs of `(key)` attribute string member and one other member.  Potentially other types, but it would give us a generic way to eat a json object that has a well specified interface without building out extra complexity in the generated code.  As you say though, it's yet another corner case, so that's unfortunate.  

Maybe it would be better to keep the schema the way google flatc supports it, but in flatcc either present the thing as a literal ubyte vector so one could parse it with another json library (I oddly like this now that I think more about it, multiple interfaces might not be great but it's _simple_ and it would cover everything with some work) or map the top level to a sorted array of string/string k/v pairs where the values are unparsed json (which would cover the string-string usecase directly but let a user parse the value if necessary). Either of those seem like they would be (comparatively) low complexity, cover the case, and in a way increase compatibility with upstream since you could even apply it to a binary flatbuffer then use a separate flexbuffer parser in the binary case.

I'm still processing what you sent me out of band but I find it very intriguing, will get back to you separately about that.

**Reply by mikkelfj** [reply]
Well, a PR that adds a (json) attribute similar to how flatcc (but not C++ I think) supports base64 or maybe hex, I don't remember, could make sense.

This would be fairly easy because there should be a precedence for other encodings, and the flatcc JSON parser is already able to skip JSON, notably it does that while looking for the type field of unions if it appears after the value field, then it reparses the JSON knowing the type.

Personally I would probably go with jquery and launch a system call or link with a library, then detect which of X syntax variants you need, and when you don't, e.g. through an enum value in schema. In this way you can transparently consume JSON and do the translation at some performance cost when necessary. You could also create you own very simple statemachine parser that passes through all text, except some detected special cases. Or use a jquery pipe as suggested earlier on.

**Reply by trws** [reply]
jquery? I only know jquery as the javascript library, but I have a feeling that's not what you mean.

**Reply by mikkelfj** [reply]
Brainfart:

https://github.com/jqlang/jq

**Reply by mikkelfj** [reply]
A quick search suggests that jq might not be fast enough for you:
https://stackoverflow.com/questions/62825963/improving-performance-when-using-jq-to-process-large-files

If you only have a few different patterns, a custom C JSON preparser might work better, as I also hinted at earlier on.

**Reply by trws** [reply]
Oh! Of course, that makes sense. We actually use jq all over the project for testing and other things (all json formatting basically) but yeah it wouldn't be fast enough in our messaging setup.  If you're curious, everything is open source over at github.com/flux-framework, mainly in the flux-core repository. We use jansson mostly, which is ergonomic and nice but not all that fast.  It would however work just fine for parsing substrings where we hit a generic json value while parsing the rest of a message with flatcc.  Is there an example or similar for a custom encoding or an element, or for getting the skipped spans while parsing a json document?

**Reply by mikkelfj** [reply]
Generic json parser.
https://github.com/dvidelabs/flatcc/blob/b09f8f4c1d41f5f6e2a66e83dbed24d380c44dca/src/runtime/json_parser.c#L639

It might not be super optimised as it is used for edge cases, but it should still be decent.
There are plenty of ways to hack a parser to much simpler and faster if you can assume the input source is either valid or will be rejected later. Note that the generated flatcc JSON parser uses a trick where it reads 8 bytes at a time, then treats it as a big endian number, in order to get early characters most significant (don't recall all details, but little endian was a bad fit). You can then shift and mask trailing bytes and compare to a known constant. In this way you can very quickly search for keywords up to 8 bytes in length. I beleive the parser also uses a trick for quickly testing if a character is present within a given 64-bit word, which is useful for looking for curly brackets while skipping space. I certainly have such code somewhere. This is not exactly SIMD level but fairly fast and portable.

Altogether, you can just create a statemachine scanning for keywords, then look for where to modify the stream, maybe create a logging array for changes, then scan over that array while copying the input buffer to output. That would be more than fast enough and easy of you don't have to do it too many times in a different manner.

**Reply by mikkelfj** [reply]
BTW: I did read about flux in detail but I have set up a slurm cluster recently and managed cluster messaging with Flatbuffers over MQTT, so does look like it is up that ballgame.

**Reply by mikkelfj** [reply]
> quickly testing if a character is present within a given 64-bit word

https://graphics.stanford.edu/~seander/bithacks.html
section: Determine if a word has a byte equal to n
